packages/driver/src/cypress/commands.js

```diff
@@ -37,6 +37,7 @@ const builtInCommands = [
   require('../cy/commands/waiting'),
   require('../cy/commands/window'),
   require('../cy/commands/xhr'),
+  require('../cy/net-stubbing').addCommand,
```

Comment:
Because you all made this internal only, I encountered a pit of snakes trying to build my own stateful commands. I never fully got to the bottom of it, but I had an `export const singleton = {}` and found that due to the way Cypress loads scripts there were two instances of the singleton being used at runtime if the singleton were used in `commands.js`, and the solution was to hackily attach state to `Cypress` object to form my singleton & just opt completely out of cypress commands & instead make pure functions that take the `cy` argument.

I'm wondering why user made plugins cannot also have state. Some users like me are writing essentially similar logic to what this PR implements as a stop gap, and finding limitations because only the "built in" plugins can access the state object.

I know doing so would allow people to break things, but I believe these plugins should be more open to extension...

Reply:
All of the APIs used in the `cy.route2` command are available to users, it doesn't use any private logic or anything like that. The `cy`, `config`, and `state` arguments passed to these `builtInCommands` handlers are available as `Cypress.cy`, `Cypress.config`, and `Cypress.state`

These APIs the corresponding types are purposefully not documented, as they are internal and subject to change at any time. You can use them if necessary but take care that new versions of Cypress may break your plugin unexpectedly.

> Some users like me are writing essentially similar logic to what this PR implements as a stop gap,

Is this source for your project somewhere? sounds interesting :)

> and finding limitations because only the "built in" plugins can access the state object.

You should be able to access `Cypress.state` fine from your plugin, though it will be untyped.

Reply:
Sorry, to clarify what I mean is I tried to make custom Cypress commands that work like the build in cy.route() stuff, where one command (cy.route) registers some state somewhere that another command (cy.server) consumes. I ran into some rough edges trying to figure that out, so thanks for the information.

I actually don't want to tamper w/ Cypress' internal state per se, I just wanted to be able to write custom cypress commands that have the same affordances like the internal commands (such as storing their own internal state).

The logic I wrote isn't open source, it was for work, but it's basically cy.server() and cy.route() for a "mock websocket transport" i attach to the window object, our application code is then decoupled to depend on a WS interface, and prefer the mock. This allows our Cypress tests to stub websockets, but the downside is we don't verify the actual WS logic anymore.  Basically we specify some pattern match that we filter the WS messages on, and a mock listener to process the WS message. I also have a console warning for any WS messages our app tries to send that the test did not mock out.

What's great about this full networking support PR is that we don't have to chose between that tradeoff, we can make our tests more "e2e" and still stub out WS messages in our in tests, so the gist of my feedback was to point out its harder to dogfood this idea with custom plugins because they're slightly different than internal plugins. I was actually diving into the source pretty deep to try to figure out how to structure my commands, but found that there were differences between writing an internal cypress command and an end user's ability to write a custom command.

Thanks again for this great testing tool, and for listening to my feedback, and providing those helpful tips!